npm_and_yarn/lib/dependabot/npm_and_yarn/package_name.rb

```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module NpmAndYarn
+    class PackageName
+      DEFINITELY_TYPED_SCOPE = /types/i.freeze
+      PACKAGE_NAME_REGEX     = %r{
+          \A                                         # beginning of string
+          (?=.{1,214}\z)                             # enforce length (1 - 214)
+          (@(?<scope>[a-z0-9\-~][a-z0-9\-\._~]*)\/)? # capture 'scope' if present
+          (?<name>[a-z0-9\-~][a-z0-9\-._~]*)         # capture package name
```

Comment:
Drive-by comment, completely ignorable: inside a character class you don't need to escape `-` if it appears first or last (e.g., `[-a-z0-9~]`) and you don't have to escape the `.` (not on this line but the one above). Just a thought to -5 characters 🙂

Also, I love the clarity of the multi-line expression with comments.